packages/flutter/lib/src/rendering/object.dart

```diff
@@ -4583,133 +4328,609 @@ mixin RelayoutWhenSystemFontsChangeMixin on RenderObject {
-  /// this fragment's children (as returned by [compileChildren]) and merge that
-  /// fragment's [config] into this fragment's [config].
+  /// Caching the semantics node ensures the id is consistent in the life time
+  /// of this object.
   ///
-  /// If a provided fragment should not merge anything into [config] call
-  /// [markAsExplicit] before passing the fragment to this method.
-  @override
-  void addAll(Iterable<_InterestingSemanticsFragment> fragments);
-
-  /// Whether this fragment wants to add any semantic information to the parent
-  /// [SemanticsNode].
-  bool get hasConfigForParent => config != null;
+  /// Note, not all `_SemanticsFragment` forms a semantics node. See
+  /// the `_SwitchableSemanticsFragment`.
+  SemanticsNode? cachedSemanticsNode;
 
-  @override
-  List<_InterestingSemanticsFragment> get mergeUpFragments => <_InterestingSemanticsFragment>[this];
+  // Use [semanticsConfiguration] to access.
+  SemanticsConfiguration? _cachedSemanticsConfiguration;
+  SemanticsConfiguration get semanticsConfiguration {
+    if (_cachedSemanticsConfiguration == null) {
+      _cachedSemanticsConfiguration = SemanticsConfiguration();
+      renderObject.describeSemanticsConfiguration(_cachedSemanticsConfiguration!);
+      assert(
+      !_cachedSemanticsConfiguration!.explicitChildNodes || _cachedSemanticsConfiguration!.childConfigurationsDelegate == null,
+      'A SemanticsConfiguration with explicitChildNode set to true cannot have a non-null childConfigsDelegate.',
+      );
+    }
+    return _cachedSemanticsConfiguration!;
+  }
 
-  Set<SemanticsTag>? _tagsForChildren;
+  // Use [getFragment] or [_getSemanticsForParent] to access.
+  _SemanticsFragment? _cachedSemanticFragment;
 
-  /// Tag all children produced by [compileChildren] with `tags`.
+  /// Updates caches stored in this object.
   ///
-  /// `tags` must not be empty.
-  void addTags(Iterable<SemanticsTag> tags) {
-    assert(tags.isNotEmpty);
-    _tagsForChildren ??= <SemanticsTag>{};
-    _tagsForChildren!.addAll(tags);
-  }
+  /// This can only be called if the owning rendering object is a semantics
+  /// boundary. For non boundary rendering objects, they require semantics
+  /// information from both their parent and child rendering objects to update
+  /// its cache, so it can't update by themselves.
+  void updateSemanticsAsBoundary() {
+    assert(semanticsConfiguration.isSemanticBoundary || renderObject.semanticsParent == null);
+    if (!kReleaseMode) {
+      FlutterTimeline.startSync('Semantics.GetFragment');
+    }
+
+    final _InterestingSemanticsFragment fragment = getFragment();
+    if (!kReleaseMode) {
+      FlutterTimeline.finishSync();
+    }
+    final List<SemanticsNode> result = <SemanticsNode>[];
+
+    if (!kReleaseMode) {
+      FlutterTimeline.startSync('Semantics.compileChildren');
+    }
+
+    fragment.compileSemanticsNodes(
+      parentSemanticsClipRect: null,
+      parentPaintClipRect: null,
```

Comment:
Aren't these rects still relevant even if we are a boundary? Why are we passing in null?